### PR TITLE
UI: Use Unification from Refinery

### PR DESCRIPTION
**Pull request by klees**
This uses the unification from `Transformation` and `Constraint` that was done in the Refinery. This leads to a **breaking change in the UI-Framework:** `withAdditionalConstraint` vanishes, because we only need `withAdditionalTransformation`.

This also includes some improvement in the Refinery (which I kindly ask @mjansenDatabay and @legionth to review):

* actually implement the unification of `Transformation` and `Constraint`
* some improvements in the traits to derive methods for `Transformation`
* a new transformation `String\StripTags`.

### CODE REVIEW

File: tests/Refinery/String/StripTagsTest.php
Comment (legionth) [inline tests/Refinery/String/StripTagsTest.php:28]: Maybe adding a test that throws the exception would be nice. Just to know that the intention is clear, that we want to work with strings here,

@mjansenDatabay  @klees  I have an additional question here too: Do you mind to add tests for both: `transform` and `applyTo`? Or do you think one test (either `transfrom` or `applyTo`) is enough?

**Review by legionth — CHANGES_REQUESTED** [review]
Awesome work :tada: 
Thank you very much! I have just a tiny remark regarding the test and also a general question below. 

I have general question to @klees and @mjansenDatabay :  Do you mind to add unit tests for both: `transform` and `applyTo`? Or do you think one test (either `transfrom` or `applyTo`) is enough?

File: tests/Refinery/String/StripTagsTest.php
Comment (mjansenDatabay) [reply tests/Refinery/String/StripTagsTest.php:28]: Good point @legionth . IMH one test, either for `transform` or `applyTo`, is sufficient. On the other hand: We currently do not force developers to call `transform` in `applyTo` or vise versa. So I suggest: If one of these methods does not call the other, a test for both should be added. IMHO we don't need to a formal rule for this.

I highly appreciate a test for the `unhappy path`.

**Comment by mjansenDatabay** [comment]
Thx for the contribution to the `Refinery` @klees . LGTM, as @legionth already mentioned a unit test for invalid `StripTags` input would be great.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
Thx for the contribution to the Refinery @klees . LGTM, as @legionth already mentioned a unit test for invalid StripTags input would be great.

File: tests/Refinery/String/StripTagsTest.php
Comment (klees) [reply tests/Refinery/String/StripTagsTest.php:28]: Hi @legionth, @mjansenDatabay!

I added the test for the sad path.

About the relation between `applyTo` and `transform`: Basically both should do the same. So if both are implemented this would be a bad sign for me. I would expect every `Transformation` to implement only one of these and derive the other via a trait. Maybe we should put this more clearly somewhere in the docu of `Refinery\Transformation`.

Thanks for looking into this!

**Review by legionth — CHANGES_REQUESTED** [review]
LGTM :shipit:

**Review by legionth — APPROVED** [review]

File: tests/Refinery/String/StripTagsTest.php
Comment (mjansenDatabay) [reply tests/Refinery/String/StripTagsTest.php:28]: @klees 

> So if both are implemented this would be a bad sign for me

+1, indeed!

> Maybe we should put this more clearly somewhere in the docu of Refinery\Transformation.

@legionth Could you please make a PR for this README addition? Thx in advance.

Comment (legionth) [reply tests/Refinery/String/StripTagsTest.php:28]: Sure

**Review by mjansenDatabay — APPROVED** [review]
Ship it!

**Comment by Amstutz** [comment]
Thanks a lot! Note that some tests are failing:

`Error: Call to undefined method ILIAS\UI\Implementation\Component\Input\Field\Duration::setAdditionalConstraint()`

Probably due to recent merge. Probably just rename to: withAdditionalTransformation

As soon as fixed, ill merge.

**Comment by klees** [comment]
@Amstutz: Done!

**Comment by Amstutz** [comment]
All good for me, just realized that this is labeled with JF due to the changes of interface. We therefore need to wait for the green light of JF before merging.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 JUL 2019 : Thanks for pointing us on this breaking change. Please merge to trunk!

**Comment by Amstutz** [comment]
Dear colleges from the refinery section, seems all good to me from the UI Components perspective, shall I do the honor and click merge?

**Comment by mjansenDatabay** [comment]
:shipit: @Amstutz  ;-).

**Comment by Amstutz** [comment]
I see that the file name of this very detective-like looking mouse is "shipit", but why? What am I missing ;-)

**Comment by mjansenDatabay** [comment]
@Amstutz https://www.quora.com/On-GitHub-what-is-the-significance-of-the-Ship-It-squirrel

**Comment by Amstutz** [comment]
@squirrel sry for calling you mouse...

**Pull request merged**
